File: NXNavigationExtensionDemo/UIKit/Feature/Drawer/DrawerAnimationController.m
//
//  DrawerAnimationController.m
//  NXNavigationExtensionDemo
//
//  Created by lidan on 2021/10/10.
//

#import "DrawerAnimationController.h"
#import "UIColor+RandomColor.h"

@interface DrawerAnimationController ()

@property (nonatomic, strong) UIVisualEffectView *backgorundColorView;

@end

@implementation DrawerAnimationController

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return 0.25;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
    UIViewController *fromVC = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIViewController *toVC = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];

    if (self.reverse) {
        [self executeReverseAnimation:transitionContext fromVC:fromVC toVC:toVC];
    } else {
        [self executeForwardsAnimation:transitionContext fromVC:fromVC toVC:toVC];
    }
}

-(void)executeForwardsAnimation:(id<UIViewControllerContextTransitioning>)transitionContext fromVC:(UIViewController *)fromVC toVC:(UIViewController *)toVC {
    UIView *containerView = [transitionContext containerView];
    [containerView insertSubview:toVC.view aboveSubview:fromVC.view];
    
    self.backgorundColorView.frame = containerView.bounds;
    self.backgorundColorView.alpha = 0.0;
    [containerView insertSubview:self.backgorundColorView belowSubview:toVC.view];
    
    CGRect toViewFinalFrame =  [transitionContext finalFrameForViewController:toVC];
    CGRect screenBounds = [[UIScreen mainScreen] bounds];
    toVC.view.frame = CGRectOffset(toViewFinalFrame, -screenBounds.size.width, 0);
    
    NSTimeInterval duration = [self transitionDuration:transitionContext];
    [UIView animateWithDuration:duration
                          delay:0.0
                        options:UIViewAnimationOptionCurveLinear
                     animations:^{
        toVC.view.frame = UIEdgeInsetsInsetRect(screenBounds, UIEdgeInsetsMake(0, 0, 0, screenBounds.size.width * 0.15));
        toVC.view.layer.cornerRadius = 10;
        toVC.view.clipsToBounds = YES;
        
        fromVC.view.transform = CGAffineTransformMakeScale(0.9, 0.9);
        fromVC.view.layer.cornerRadius = 10;
        fromVC.view.clipsToBounds = YES;
        
        self.backgorundColorView.alpha = 1.0;
    } completion:^(BOOL finished) {
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
}

-(void)executeReverseAnimation:(id<UIViewControllerContextTransitioning>)transitionContext fromVC:(UIViewController *)fromVC toVC:(UIViewController *)toVC {
    // fix: white/black screen
    if (fromVC.modalPresentationStyle == UIModalPresentationFullScreen) {
        UIView *containerView = [transitionContext containerView];
        [containerView insertSubview:toVC.view belowSubview:fromVC.view];
    }
    
    CGRect fromViewFinalFrame =  [transitionContext finalFrameForViewController:fromVC];
    
    NSTimeInterval duration = [self transitionDuration:transitionContext];
    [UIView animateWithDuration:duration
                          delay:0.0
                        options:UIViewAnimationOptionCurveLinear
                     animations:^{
        toVC.view.layer.cornerRadius = 0;
        toVC.view.clipsToBounds = NO;
        toVC.view.transform = CGAffineTransformIdentity;
        
        fromVC.view.frame = CGRectOffset(fromViewFinalFrame, -fromViewFinalFrame.size.width, 0);
        
        self.backgorundColorView.alpha = 0.0;
    } completion:^(BOOL finished) {
        self.reverse = NO;
        if ([transitionContext transitionWasCancelled]) {
            toVC.view.layer.cornerRadius = 0;
            toVC.view.clipsToBounds = NO;
            toVC.view.transform = CGAffineTransformIdentity;
            
            self.backgorundColorView.alpha = 1.0;
        } else {
            [self.backgorundColorView removeFromSuperview];
        }
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
}

#pragma mark - Getter

- (UIVisualEffectView *)backgorundColorView {
    if (!_backgorundColorView) {
        UIBlurEffect *effect;
        if (@available(iOS 13.0, *)) {
            effect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemChromeMaterial];
        } else {
            effect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleExtraLight];
        }
        
        _backgorundColorView = [[UIVisualEffectView alloc] initWithEffect:effect];
        _backgorundColorView.alpha = 0.0;
    }
    return _backgorundColorView;
}

@end

File: Examples/Shared/UIKit/Feature/Drawer/DrawerAnimationController.swift
//
//  DrawerAnimationController.swift
//  NXNavigationExtensionDemo
//
//  Created by lidan on 2021/11/9.
//

import UIKit

class DrawerAnimationController: NSObject {
    
    private lazy var backgroundColorView: UIVisualEffectView = {
        var effect = UIBlurEffect(style: .extraLight)
        if #available(iOS 13.0, *) {
            effect = UIBlurEffect(style: .systemChromeMaterial)
        }
        
        let backgroundColorView = UIVisualEffectView(effect: effect)
        backgroundColorView.alpha = 0.0
        return backgroundColorView
    }()

    var isReverse: Bool = false
    
}

extension DrawerAnimationController {
    
    private func executeForwardsAnimation(using transitionContext: UIViewControllerContextTransitioning?, fromVC: UIViewController, toVC: UIViewController) {
        let containerView = transitionContext?.containerView
        containerView?.insertSubview(toVC.view, aboveSubview: fromVC.view)
        
        backgroundColorView.frame = containerView?.bounds ?? .zero
        backgroundColorView.alpha = 0.0
        containerView?.insertSubview(backgroundColorView, belowSubview: toVC.view)
        
        let toViewFinalFrame = transitionContext?.finalFrame(for: toVC) ?? .zero
        let screenBounds = fromVC.view.frame
        toVC.view.frame = toViewFinalFrame.offsetBy(dx: -screenBounds.width, dy: 0)
        
        let duration = transitionDuration(using: transitionContext)
        UIView.animate(withDuration: duration, delay: 0.0, options: .curveLinear) { [weak self] in
            toVC.view.frame = screenBounds.inset(by: UIEdgeInsets(top: 0, left: 0, bottom: 0, right: screenBounds.width * 0.1))
            fromVC.view.transform = CGAffineTransform(scaleX: 0.9, y: 0.9)
            self?.backgroundColorView.alpha = 1.0
        } completion: { finished in
            transitionContext?.completeTransition(!(transitionContext?.transitionWasCancelled ?? true))
        }
    }
    
    private func executeReverseAnimation(using transitionContext: UIViewControllerContextTransitioning?, fromVC: UIViewController, toVC: UIViewController) {
        // fix: white/black screen
        if fromVC.modalPresentationStyle == .fullScreen {
            let containerView = transitionContext?.containerView
            containerView?.insertSubview(toVC.view, belowSubview: fromVC.view)
        }
        
        let fromViewFinalFrame = transitionContext?.finalFrame(for: fromVC) ?? .zero
        let duration = transitionDuration(using: transitionContext)
        
        UIView.animate(withDuration: duration, delay: 0.0, options: .curveLinear) { [weak self] in
            toVC.view.transform = .identity
            
            fromVC.view.frame = fromViewFinalFrame.offsetBy(dx: -fromViewFinalFrame.width, dy: 0)
            
            self?.backgroundColorView.alpha = 0.0
        } completion: { [weak self] finished in
            self?.isReverse = false
            if transitionContext?.transitionWasCancelled ?? true {
                toVC.view.transform = .identity
                
                self?.backgroundColorView.alpha = 1.0
            } else {
                self?.backgroundColorView.alpha = 1.0
            }
            transitionContext?.completeTransition(!(transitionContext?.transitionWasCancelled ?? true))
        }
    }
    
}

extension DrawerAnimationController: UIViewControllerAnimatedTransitioning {
    
    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return 0.25
    }
    
    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        guard let fromVC = transitionContext.viewController(forKey: .from),
              let toVC = transitionContext.viewController(forKey: .to) else { return }
        
        if isReverse {
            executeReverseAnimation(using: transitionContext, fromVC: fromVC, toVC: toVC)
        } else {
            executeForwardsAnimation(using: transitionContext, fromVC: fromVC, toVC: toVC)
        }
    }
    
}
